zip-0308.rst

```diff
@@ -0,0 +1,365 @@
+
+These have been chosen by guesswork. Should we change any of them?
+
+In particular, if the amount to migrate is large, then this strategy can
+result in large amounts (up to 990 ZEC, worth USD ~49500 at time of writing)
+transferred in each transaction. This leaks the fact that the transaction was
+sent by a user who has at least that amount.
+
+The strategy does not migrate any remaining fractional amount less than
+0.01 ZEC (worth USD ~0.50 at time of writing). Is this reasonable?
+
+In deciding the amount to send in each transaction, the strategy does not
+take  account of the values of individual Sprout notes, only the total amount
+remaining to migrate. Can a strategy that is sensitive to individual note
+values improve privacy?
+
+An adversary may attempt to interfere with the view of the block chain seen
+by a subset of nodes that are performing migrations, in order to cause those
+nodes to send migration batches at a different time, so that they may be
+distinguished. Is there anything further we can do to mitigate this
+vulnerability?
+
+
+RPC calls
+---------
+
+Nodes MUST maintain a boolean state variable during their execution, to
+determine whether migration is enabled. The default when a node starts, is
+set by a configuration option::
+
+  -migration=0/1
+
+The destination z-address can optionally be set by another option::
+
+  -migrationdestaddress=<zaddr>
+
+If this option is not present then the migration destination address is
+the address for Sapling account 0, with the default diversifier [#zip-0032]_.
+
+The state variable can also be set for a running node using the following
+RPC method::
+
+  setmigration true/false
+
+It is intentional that the only option associated with the migration is the
+destination z-address. Other options could potentially distinguish users.
+
+
+Nodes MUST also support the following RPC call to return the current status of
+the migration::
+
+  getmigrationstatus
+
+Returns::
+
+  {
+    "enabled": true|false,
+    "destination_address": "zaddr",
+    "unmigrated_amount": nnn.n,
+    "unconfirmed_migrated_amount": nnn.n,
+    "confirmed_migrated_amount": nnn.n,
+    "confirmed_migration_transactions": nnn,
+    "time_started": ttt, // Unix timestamp
+    "migration_txids": [txids]
+  }
+
+The ``destination_address`` field MAY be omitted if the ``-migrationaddress``
+parameter is not set and no default address has yet been generated.
+
+The values of ``unmigrated_amount`` and ``migrated_amount`` MUST take into
+account failed transactions, that were not mined within their expiration
+height.
+
+``txids`` is a list of strings representing transaction IDs of all known
+migration transactions involving this wallet, as lowercase hexadecimal in
+RPC byte order. A given transaction is defined as a migration transaction
+iff it has:
+ * one or more Sprout JoinSplits with nonzero vpub_new field; and
+ * no Sapling Spends, and;
+ * one or more Sapling Outputs.
```

Comment:
When would a migration transaction have more than one Sapling output?

Reply:
A migration transaction generated by an implementation of this spec would not. However, I'm not sure it makes sense to exclude migration transactions that could have been generated manually.

Reply:
Added a note "it is possible that manually created transactions involving this wallet will be recognized as migration transactions and included in ``txids``."